geth/jail/internal/loop/loop.go

```diff
@@ -40,6 +64,7 @@ type Loop struct {
 	lock  sync.RWMutex
 	tasks map[int64]Task
 	ready chan Task
+	state
```

Comment:
I don't see any benefit in having second mutex for the boolean variable holding `state`. Both `isAccepting` and `close` are private methods and calls to them are already guarded by `lock` mutex. For example in `(*loop).close()` method `lock` already guarantees that nobody will touch `state`, and we have two redundant (and not very cheap) lock/unlock indirect calls.

How would this solution look like if `state` was simple bool field of `Loop` type?

Reply:
My original pull request had your suggested implementation, but @JekaMas thought it was a bad idea. He was concerned that errors might be made down the road because contributors must remember that the boolean should be protected by the mutex. He suggested hiding the mutex logic when getting and setting the boolean.

Also, an issue that I ran into by locking the mutex in `loop.Ready` (which would need to happen to ensure we do not send on a closed channel) was that `loop.Run` would need to be changed so that `loop.processTask` ran in its own goroutine. If not, then we would get a deadlock when `loop.removeByID` tried to acquire the lock and `loop.Ready` was trying to add another task to an unbuffered `loop.Ready`.

I would be happy to switch it back, but I am getting conflicting suggestions.

Reply:
@thedadams first of all, sorry for the confusion and conflicting suggestions. I found original comments with suggestions and I disagree with them.

As for the `loop.Ready`, it's fine because `(*loop).close()` is called only when `(*loop).Run` is finished, so there is really not that much sense of guarding sending into the channel with mutex in `(*loop).Ready`. Even more, state of the channel may be used an indicator of state (loop open/closed), but as this PR adds boolean state variable, there not that much sense in closing `l.ready` channel at all.

Let's do the following way — I leave it to you to decide what to do with `state`, ditch it in favor of simpler code and your original approach or leave it as is. I'm strongly for the first option, but this PR takes too much time already, and `loop` package needs revisiting and refactoring anyway, so I think both option would work for now.

Reply:
The original issue stated that `(*loop).Ready` should return an error if a task is readied on a loop that is closed. So at the very least we need to look at the boolean in the loop to see if the loop is accepting tasks. This would mean we at least call `(*loop).lock.RLock()` to look at this boolean. That would create the same issue where `(*loop).processTask` would need to be called in its own goroutine.